control-plane/agents/src/bin/core/controller/io_engine/mod.rs

```diff
@@ -135,6 +136,12 @@ where
     }
 }
 
+#[async_trait]
+pub(crate) trait SnapshotApi {
+    /// Create a nexus snapshot via gRPC.
+    async fn create_snapshot(&self, request: &dyn SnapshotRequest) -> Result<SnapshotId, SvcError>;
```

Comment:
I don't think this trait here helps, because replica and nexus snapshots have different requirements, also we use associated types for this module:

```suggestion
    type Create = Send + Sync;
    type CreateOutput = Send + Sync + Sized;
    /// Create a nexus snapshot via gRPC.
    async fn create_snapshot(&self, request: &Self::Create) -> Result<Self::CreateOutput, SvcError>;
```

Reply:
If I understand your comment right, are you suggesting like this?

1. The types Nexus and Replica need to impl SnapshotApi trait.
2. `impl SnapshotApi for Nexus` will set type `Create = CreateNexusSnapshot`, and `CreateOutput = SnapshotId`.
3. `impl SnapshotApi for Replica` will set type `Create = CreateReplicaSnapshot`, and `CreateOutput = SnapshotId`.

Reply:
Yes, though the output will not just be SnapshotId, it will be their respective outputs, so:

1. The types Nexus and Replica need to impl SnapshotApi trait.
2. `impl SnapshotApi for Nexus` will set type `Create = CreateNexusSnapshot`, and `CreateOutput = NexusSnapshotResp`.
3. `impl SnapshotApi for Replica` will set type `Create = CreateReplicaSnapshot`, and `CreateOutput = ReplicaSnapshotResp`.